File: lib/features/autocomplete-plus.js
'use babel'

import SubAtom from 'sub-atom'
import loglevel from 'loglevel'

const log = loglevel.getLogger('ensime.autocomplete-plus-provider')

export default function(clientLookup) {
  const disposables = new SubAtom()
  let noOfAutocompleteSuggestions

  disposables.add(
    atom.config.observe('Ensime.noOfAutocompleteSuggestions', value => {
      noOfAutocompleteSuggestions = value
    })
  )

  return {
    dispose() {
      return disposables.dispose()
    },
    getCompletions(textBuffer, bufferPosition, callback) {
      const { formatCompletionsSignature } = require('../atom-formatting')
      const file = textBuffer.getPath()
      const offset = textBuffer.characterIndexForPosition(bufferPosition)
      return __guard__(clientLookup(textBuffer), x =>
        x
          .getCompletions(
            file,
            textBuffer.getText(),
            offset,
            noOfAutocompleteSuggestions
          )
          .then(function(result) {
            let c
            const { completions } = result

            if (completions) {
              const translate = function(c) {
                const { typeSig } = c
                if (c.isCallable) {
                  const formattedSignature = formatCompletionsSignature(
                    typeSig.sections
                  )
                  return {
                    leftLabel: c.typeSig.result,
                    snippet: `${c.name}${formattedSignature}`,
                    fullyQualified: c.typeSig.result
                  }
                } else {
                  return {
                    snippet: c.name
                  }
                }
              }

              const autocompletions = (() => {
                const result1 = []
                for (c of Array.from(completions)) {
                  result1.push(translate(c))
                }
                return result1
              })()
              return callback(autocompletions)
            }
          })
      )
    },

    onDidInsertSuggestion({ editor, suggestion, triggerPosition }) {
      return log.trace(['inserted suggestion', suggestion])
    }
  }
}

function __guard__(value, transform) {
  return typeof value !== 'undefined' && value !== null
    ? transform(value)
    : undefined
}

File: src/main/scala/features/autocomplete-plus.scala
package org.ensime.atom.features

import scala.scalajs.js
import scala.scalajs.js.annotation._

import org.ensime.atom.AtomFormatting
import org.ensime.atom.external.{
  Api,
  AtomEnvironment,
  Completion,
  LogLevel,
  Point,
  SubAtom,
  TextBuffer,
  TextEditor
}

@JSExportAll
@JSExportTopLevel("AutocompletePlusProvider")
class AutocompletePlusProvider(clientLookup: js.Function1[TextBuffer, Api]) {

  val log = LogLevel.getLogger("ensime.autocomplete-plus-provider")
  val disposables = new SubAtom()
  var noOfAutocompleteSuggestions: js.UndefOr[Int] = js.undefined;

  disposables.add(AtomEnvironment.config.observe[Int]("Ensime.noOfAutocompleteSuggestions", value => {
    noOfAutocompleteSuggestions = value
  }))

  def dispose(): Unit = disposables.dispose()

  def getCompletions(textBuffer: TextBuffer, bufferPosition: Point, callback: js.Function1[js.Array[js.Dynamic], _]): Unit = {
    val file = textBuffer.getPath()
    val offset = textBuffer.characterIndexForPosition(bufferPosition)
    clientLookup(textBuffer).getCompletions(file, textBuffer.getText(), offset, noOfAutocompleteSuggestions).andThen((result) => {
      result.completions.foreach{ completions =>
        def translate(completion: Completion): js.Dynamic = {
          completion.typeInfo.map{typeInfo =>
            this.dispose
            val formattedSignature = AtomFormatting.formatCompletionsSignature(typeInfo.typeArgs)
            js.Dynamic.literal(
              leftLabel = typeInfo.name,
              snippet = s"${completion.name}${formattedSignature}",
              fullyQualified = typeInfo.fullName,
            )
          }.getOrElse{
            js.Dynamic.literal(
              snippet = completion.name
            )
          }
        }
        this.dispose
        val autocompletions = completions.map(translate _)
        callback(autocompletions)
      }
    });
  }

  def onDidInsertSuggestion(arg: OnDidInsertSuggestionArg ): Unit =
    log.trace(js.Array("inserted suggestion", arg.suggestion))
}

@js.native
trait OnDidInsertSuggestionArg extends js.Object {
  val editor: TextEditor = js.native
  val suggestion: js.Any = js.native
  val triggerPosition: String = js.native
}
